packages/editor/src/editor.ts

```diff
@@ -0,0 +1,236 @@
+  _instanceState: ShellInstanceState;
+  _makeMultilineJSIntoSingleLine: any;
+  _loadExternalCode: any;
+  require: any;
+  _editorSession: EditorSession;
+  print: (...args: any[]) => Promise<void>;
+
+  constructor({ input, vscodeDir, tmpDir, instanceState, makeMultilineJSIntoSingleLine, loadExternalCode }: EditorOptions) {
+    this._input = input;
+    this._vscodeDir = vscodeDir;
+    this._tmpDir = tmpDir;
+    this._instanceState = instanceState;
+    this._makeMultilineJSIntoSingleLine = makeMultilineJSIntoSingleLine;
+    this._loadExternalCode = loadExternalCode;
+    this._editorSession = { cmd: '', args: [], content: '', code: '', tmpDoc: '' };
+    this.print = instanceState.context.print;
+
+    // Add edit command support to shell api.
+    const wrapperFn = (...args: string[]) => {
+      return Object.assign(this.runEditCommand(args), {
+        [Symbol.for('@@mongosh.syntheticPromise')]: true
+      });
+    };
+
+    wrapperFn.isDirectShellCommand = true;
+    wrapperFn.returnsPromise = true;
+    (instanceState.shellApi as any).edit = instanceState.context.edit = wrapperFn;
+    (signatures.ShellApi.attributes as any).edit = {
+      type: 'function',
+      returnsPromise: true,
+      isDirectShellCommand: true
+    } as TypeSignature;
+  }
+
+  activate(): void {
+    this._instanceState.registerPlugin(this);
+  }
+
+  // In the case of using VSCode as an external editor,
+  // we should detect whether the MongoDB extension is installed and open a .mongodb file.
+  // In all other cases we should open a .js file.
+  async _getExtension(cmd: string): Promise<string> {
+    if (!this._isVscodeApp(cmd)) {
+      return 'js';
+    }
+
+    try {
+      const extensions = await fse.readdir(path.join(this._vscodeDir, 'extensions'));
+      const hasMongodbExtension = !!extensions
+        .find((name) => name.includes('mongodb.mongodb-vscode'));
+
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-done', {
+        vscodeDir: this._vscodeDir,
+        hasMongodbExtension
+      });
+
+      return hasMongodbExtension ? 'mongodb' : 'js';
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-failed', {
+        action: 'mongosh-editor:read-vscode-extensions-failed',
+        error: error.message
+      });
+
+      return 'js';
+    }
+  }
+
+  async _getEditor(): Promise<string|null> {
+    // Check for an external editor in the mongosh configuration.
+    let editor: string | null = await this._instanceState.shellApi.config.get('editor');
+
+    // Check for an external editor in the environment variable.
+    if (!editor && process.env.EDITOR) {
+      editor = process.env.EDITOR;
+    }
+
+    return editor;
+  }
+
+  async _createTempFile(): Promise<void> {
+    const ext = await this._getExtension(this._editorSession.cmd);
+    const content = await this._getEditorContent();
+
+    this._editorSession.tmpDoc = path.join(this._tmpDir, `edit-${uuidv4()}.${ext}`);
```

Comment:
Regarding `uuidv4()` – in mongosh, we generally use `ObjectId()`s to generate unique filenames, including for the logs (which has a few other advantages, like being able to tell the creation time or identify which files come from the same process). If there’s a reason to use UUID v4 specifically, then the `bson` package also provides an implementation of that, so I think using that would be a bit more convenient given that we already use `bson` across the board :)